xds/internal/xdsclient/client_new.go

```diff
@@ -152,8 +157,7 @@ func NewForTesting(opts OptionsForTesting) (XDSClient, func(), error) {
 	if err := bootstrap.SetFallbackBootstrapConfig(opts.Contents); err != nil {
 		return nil, nil, err
 	}
-	client, cancel, err := newRefCounted(opts.Name, opts.WatchExpiryTimeout, opts.IdleChannelExpiryTimeout, opts.StreamBackoffAfterFailure)
-	return client, func() { bootstrap.UnsetFallbackBootstrapConfigForTesting(); cancel() }, err
+	return newRefCounted(opts.Name, opts.WatchExpiryTimeout, opts.IdleChannelExpiryTimeout, opts.StreamBackoffAfterFailure)
```

Comment:
I feel that it's better to have tests clean up their own mutations/side-effects. This would prevent issues in which test fail only when run in a certain order.  Since the `BootstrapContentsForTesting` ServerOption is public, adding a new param to it is also not preferable. I don't know the best way to do this 😕.

Reply:
I had thought about making callers of `NewForTesting` ensure that bootstrap configuration is set either by setting one of the associated env vars or by calling `SetFallbackBootstrapConfig`. If we did that, we can remove the `Contents` field from `OptionsForTesting` and have the tests setup the bootstrap configuration and do the cleanup themselves. But the number of callsites for `NewForTesting` is enormous. So, I didn't want to do that as part of this PR.

But I agree with your concern, and maybe I can do that as a follow-up PR so that we take care of this flake first, and then cleanup the remaining tests. 

What do you think?

Reply:
I have to agree with Arjan here. Ever since I joined the team 3.5 years ago, Doug has always stressed to not set any globals amongst tests that persist over test iterations that would couple tests in any way.

Any time I have done this, he has always made me rewrite the test in order to be hermetic and not coupled with other tests. I agree with this because of the ordering thing Arjan mentioned, I think go test guarantees things are run serially but I forgot what the strict ordering requirements are, and also it's weird to have one test state affect another test state.

But since test is blocking development flow I'm fine merging this now and following up on it to address this issue.

Reply:
I'm fine with doing the cleanup in a later PR.

Reply:
Ack. Thanks.